doc/dev/cephfs-mirroring.rst

```diff
@@ -0,0 +1,78 @@
+snapshot data to the remote filesystem and creating a snapshot with the same name as the
+snapshot being synchronized.
+
+Change Detection
+################
+
+The first snapshot to get synchronized involves bulk transfer of data to the remote
+filesystem. Subsequent snapshot synchronization transfers changed/modified files.
+Identifying changed files is based on `Efficient Tree Traversal` algorithm detailed
+here :doc:`/dev/cephfs-mirroring-tree-traversal`.
+
+Identifying changed files between two snapshots (one of which is already synchronized),
+involves (efficiently) walking two (snapshot) trees. Since one of the snapshots is
+already synchronized, scanning the two snapshots locally would suffice. For the most
+part, identifying changed files between two snapshot does not involve querying the
+remote filesystem. However, there are cases where it is essential to query the remote
+filesystem:
+
+Failed Synchronization
+**********************
+
+Failure to transfer snapshot data (which could be due to variety of reasons) leaves
+the remote filesystem with incomplete data. Although retrying synchronization most
+likely overcomes transient issues, non-recoverable errors would need to be handled.
+When a snapshot synchronization hits a non-recoverable error, the snapshot is noted
+for failed synchronization and gets skipped. This is done either manually by the
+operator or when the number of consecutive retries (post failure) hits an configured
+limit.
+
+Skipping a "failed" snapshot leaves incomplete data on the remote filesystem. When the
+next snapshot is chosen for synchronization, local scanning of (two) snapshots is not
+feasible, since certain files (which were part of the "failed" snapshot) may have been
+transferred to the remote filesystem but do not exist in the "next" chosen snapshot.
+In such as case, scanning the two snapshots "locally" for changes would ignore the
+incomplete data on the remote filesystem left behind. In such as case identifying
+changes would need to compare files between the snapshot and the remote filesystem.
+Alternatively, the remote filesystem data could be purged followed by bulk transfer of
+data from the next chosen snapshot. Subsequent snapshots can then rely on scanning local
+snaphots for identifying changes to be transferred.
+
+Deleted Snapshots
+*****************
+
+Snapshots on the primary filesystem can be deleted even during synchronization.
+This needs to be handled by the mirror daemon and poses the same issues as mention
+above. Moreover, mirroring can be disabled in midst of snapshots being synchronized
+and enabled after deleting snapshots (including the one under synchronization) and
+poses the same set of issues.
+
+Snapshot Synchronization Order
+##############################
+
+Although the order in which snapshots get chosen for synhronization does not matter,
+it may be beneficial to pick snapshots based on creation order (using snap-id).
+
+Snapshot Incarnation
```

Comment:
This would mean that the snap-replicator code needs to test for the presence of the new snapshot before embarking on the actual replication to ascertain if the snap creation with the new name at the end of the replication will be successful for not.
If the new snapshot already exists, then it would need to be deleted first.
Would the `client.admin` role or anything appropriate ever have snap creation authority but not snap delete authority ?
This authorization would be critical to implementing the feature.

Reply:
> This would mean that the snap-replicator code needs to test for the presence of the new snapshot before embarking on the actual replication to ascertain if the snap creation with the new name at the end of the replication will be successful for not.

Kind of. however, the snap "cleanup" would be done before kicking off the replicator to start mirroring a snapshot.

> If the new snapshot already exists, then it would need to be deleted first.
> Would the `client.admin` role or anything appropriate ever have snap creation authority but not snap delete authority ?
> This authorization would be critical to implementing the feature.

I don't get the point about "client.admin"? Do you mean the user for the remote cluster? The remote cluster user should have the capability to create and delete snapshots ("s" flag).